### PR TITLE
Migrate Qwen3 to NNX

### PR DESCRIPTION
**Pull request by shuningjin**
# Description

1. migrate Qwen3DecoderLayer (dense) and Qwen3MoeDecoderLayer (moe) from Linen to NNX. 
- shared component: RMSNorm + Attention + RMSNorm
- the dense model is close to llama2 (MlpBlock), https://github.com/AI-Hypercomputer/maxtext/pull/2370
- the moe model is close to mixtral (RoutedMOE), https://github.com/AI-Hypercomputer/maxtext/pull/2166
2. change variable name to be consistent with other decoder [layers](https://github.com/AI-Hypercomputer/maxtext/tree/main/src/MaxText/layers) (e.g., llama2, mixtral, deepseek)
3. axis: activation_length -> activation_norm_length. 
- activation_norm_length is used in other decoder layers, and is the correct way. Fix: b/441547754

# Tests

Details in b/448171914

## [dense: qwen3-4b]

train: v5p-8
- cmd: https://paste.googleplex.com/6224473325961216 
- https://diff.googleplex.com/#key=S714ns5bS1fi
- memory same, loss / tflop similar

decode: v6e-8
- cmd: https://paste.googleplex.com/5372041654042624
- https://diff.googleplex.com/#key=DgewTDqTEv37
- Memstats same, RAMstats slight diff (24.04 vs. 23.98), output same

jetstream: v6e-8
- log1: 
  - cmd: https://paste.googleplex.com/6306253127155712
  - https://diff.googleplex.com/#key=BSHtWmusqNyQ
  - memstats same, RAMstats slight diff (23.29 vs. 23.69)
- log2: 
  - cmd: https://paste.googleplex.com/6028629595258880
  - https://diff.googleplex.com/#key=e02j0rUAX2QD
  - acc same 0.6942
- profile 
  - cmd: https://paste.googleplex.com/4892507330707456
  - before (6s): https://xprof.corp.google.com/memory_profile/shuningjin-13138045314128610787
  - after (6s): https://xprof.corp.google.com/memory_profile/shuningjin-5512421363322943186
  - similar peak mem ~3.89GiB, id match

## [moe: qwen3-30b-a3b]

train: v5p-8
- cmd: https://paste.googleplex.com/4815163307982848
- https://diff.googleplex.com/#key=IfEi6RXU05u9
- memory same, loss / tflop similar

decode: v6e-8
- cmd: https://paste.googleplex.com/4808862590959616
- https://diff.googleplex.com/#key=wlOtFS9GpTMD
- Memstats same, RAMstats slight diff (56.31 vs. 56.82), output same

jetstream: v6e-8
- log1: 
  - cmd: https://paste.googleplex.com/6027123068370944
  - https://diff.googleplex.com/#key=bTprvexIVOLd
  - same memstats, RAMstats slight diff (60.57 vs. 60.93)
- log2: 
  - cmd: https://paste.googleplex.com/6320961863417856
  - https://diff.googleplex.com/#key=UzGxEatlvgLW
  - same acc=0.792
- profile (ignore this, new profile with 6s in comment)
  - cmd: https://paste.googleplex.com/4850704028139520
  - before (1s): https://xprof.corp.google.com/memory_profile/shuningjin-10823727747969299875
  - after (1s): https://xprof.corp.google.com/memory_profile/shuningjin-15805069810789941347
  - similar peak mem ~11.40GiB, id match

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

**Comment by RissyRan** [comment]
Not sure why it didn't work. Logs seems fine to me: https://github.com/AI-Hypercomputer/maxtext/actions/runs/18202874722/job/51825960157

**Review by RissyRan — APPROVED** [review]
LGTM

**Comment by shuningjin** [comment]
> For the MoE version, I see the same issue @RissyRan has in #2430. Some of the modules are missing in the after profile. But the HLOs match for the ones that are there. Could this be tied to the window the profiler actually ran for?

Maybe it is related to the profiling duration. For dense version, I capture 6000ms (6s). For MoE version, I only capture 1000ms (1s). I use the same duration for both before/after.

I see https://github.com/AI-Hypercomputer/maxtext/pull/2430 also capture 1000ms (1s): https://screenshot.googleplex.com/9RWK6eMaWRyKonv

**Comment by shuningjin** [comment]
I re-capture the qwen3-30b-a3b (moe) profile with longer profiling duration=6s. Now, before and after have same number of modules (16). And the HLO ids still match. @bvandermoon 

before (6s): https://xprof.corp.google.com/memory_profile/shuningjin-3133062836637372009
after (6s): https://xprof.corp.google.com/memory_profile/shuningjin-18001255487583733129

File: src/MaxText/layers/qwen3.py [hunk 1]
```diff
@@ -21,111 +21,151 @@
...
+  def apply_attention_with_norm(
+      self,
+      inputs: jnp.ndarray,
+      decoder_segment_ids: None | jnp.ndarray,
+      decoder_positions: None | jnp.ndarray,
+      deterministic: bool,
+      model_mode: str,
+  ):
+    """Applies self-attention with pre and post-layer normalization."""
+    inputs = nn.with_logical_constraint(inputs, self.activation_axis_names)
```
Comment (bvandermoon) [inline src/MaxText/layers/qwen3.py:118]: Seems this one was not here before?

File: src/MaxText/layers/qwen3.py [hunk 2]
```diff
@@ -137,40 +177,17 @@ def __call__(
...
-    )(hidden_states, deterministic=deterministic)
-
-    # Final residual connection
-    layer_output = residual_after_attention + mlp_output
-    layer_output = nn.with_logical_constraint(
-        layer_output,
-        ("activation_batch", "activation_length", "activation_embed"),
-    )
+    mlp_lnx = self.mlp(hidden_states, deterministic=deterministic)
+    mlp_lnx = nn.with_logical_constraint(mlp_lnx, self.activation_axis_names)
```
Comment (bvandermoon) [inline src/MaxText/layers/qwen3.py:185]: Looks like this one was not here previously

**Review by bvandermoon — APPROVED** [review]
LGTM, just a couple questions on newly added `nn.with_logical_constraint()`. Please confirm those are added intentionally

File: src/MaxText/layers/qwen3.py (hunk 1 above)
Comment (shuningjin) [reply src/MaxText/layers/qwen3.py:118]: Thanks for checking! This is to be consistent with [llama2](https://github.com/AI-Hypercomputer/maxtext/blob/698fa045be60bba2f0f4b64d85c34a04c223d2a6/src/MaxText/layers/llama2.py#L143) and [mixtral](https://github.com/AI-Hypercomputer/maxtext/blob/4a261bf818d779f0d3661daf8d505ab5b3d9ab80/src/MaxText/layers/mixtral.py#L63).

File: src/MaxText/layers/qwen3.py (hunk 2 above)
Comment (shuningjin) [reply src/MaxText/layers/qwen3.py:185]: This is to be consistent with [llama2](https://github.com/AI-Hypercomputer/maxtext/blob/698fa045be60bba2f0f4b64d85c34a04c223d2a6/src/MaxText/layers/llama2.py#L171) and [mixtral](https://github.com/AI-Hypercomputer/maxtext/blob/4a261bf818d779f0d3661daf8d505ab5b3d9ab80/src/MaxText/layers/mixtral.py#L148).
